File: nim/allergies/allergies.nim
type
  Allergen* = enum
    Eggs,
    Peanuts, 
    Shellfish, 
    Strawberries, 
    Tomatoes, 
    Chocolate, 
    Pollen, 
    Cats

  
proc allergies*(score: int): set[Allergen] =
  cast[set[Allergen]](score)

proc isAllergicTo*(score: int, allergen: Allergen): bool =
  allergen in allergies(score)

File: ocaml/allergies/allergies.ml
open Base

type allergen = Eggs
              | Peanuts
              | Shellfish
              | Strawberries
              | Tomatoes
              | Chocolate
              | Pollen
              | Cats

let allergies =
    [Eggs; Peanuts; Shellfish; Strawberries; Tomatoes; Chocolate; Pollen; Cats]

let to_score = function
    | Eggs -> 1
    | Peanuts -> 2
    | Shellfish -> 4
    | Strawberries -> 8
    | Tomatoes -> 16
    | Chocolate -> 32
    | Pollen -> 64
    | Cats -> 128

let allergic_to score allergen =
    let s = to_score allergen in score land s = s

let allergies score =
    List.filter allergies ~f:(allergic_to score)
    
